### ISSUE TITLE
Fasta: Generalized void

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#30470]
This is the Fasta specific issue for issue #30176, which contains the details.

**Comment by anders-sandholm** [comment]
From talking with @lrhn it sounds like this will block the next wave of library changes. 

@kmillikin, @stefantsov, just wondering if there is a way we could/should move this forward in time.

**Comment by lrhn** [comment]
To be more specific, the next wave will include changes to asynchronous classes like `Future` and `Stream`. Some methods currently return `Future` where they should return `Future<void>`. We'd like to make those changes at the same time as all other changes to those the classes. In practice, it's a small change since `void` and `dynamic` are equivalent wrt. subtyping, so `Future<void>` and `Future<dynamic>` are too, but it would be great to have it early so we can see if some use-places are affected (you'd not be allowed to use a value with static type `void`, you would when it has static type `dynamic`).

Some methods return `Future<Null>` where they should return `Future<void>`. Those changes are bigger changes - it's an error to change a subclass before the superclass/interface has been updated because `Future<void>` is not a subtype of `Future<Null>`, it's the other way around.
Luckily we don't have any of those in the public parts of the platform libraries.

**Comment by peter-ahe-google** [comment]
Do we have any tests of this?

**Comment by lrhn** [comment]
A quick search shows `language_2/generalized_void_syntax_test.dart`, `language_2/void_type_callbacks_test.dart`, `language_2/void_type_override_test.dart` and `language_2/void_type_usage_test.dart` as containing `<void>`.

**Comment by peter-ahe-google** [comment]
I need a list of the actual tests that the language teams has implemented to test this feature. I know how to do a quick search.

**Comment by peter-ahe-google** [comment]
Proposed fix: [change 30760](https://dart-review.googlesource.com/30760).

**Comment by eernstg** [comment]
I added a positive test here: https://dart-review.googlesource.com/c/sdk/+/30840. It's simply checking that expressions of type `void` are allowed to occur in the situations where we want them to be allowed (I ran the test with many configurations and tools, and none of them failed. Hurrah! ;-).

**Comment by eernstg** [comment]
NB: [Change 30760](https://dart-review.googlesource.com/30760) allows the `void` type to occur as a type, but it does not ensure, e.g., that the following compile-time error occurs:
```
void get x => null;

void main() {
  var y = x; //# 01: compile-time error
}
```
So it would be a fix for part of this issue, not all of it. But a partial fix is also useful, of course!

**Comment by mraleph** [comment]
Note: this is important for being able to run Flutter without any issues. That's why we are bumping priority.

For the record there is a patch from @peter-ahe-google https://dart-review.googlesource.com/c/sdk/+/30760

**Comment by peter-ahe-google** [comment]
I'm still short on tests of the new features, I do have tests of existing features that have been in Dart for years.

**Comment by eernstg** [comment]
Here are some tests in `sdk/tests/language_2` that contain constructs using `void` in a way that used to be unsupported, and whose expectations match the semantics when the 'generalized void' feature is implemented. In other words, they should run successfully when void is generalized, and they do test things that are relevant for the generalized void feature.

- `generalized_void_syntax_test.dart`: This test uses uses `void` in positions that used to be compile-time errors (typically: syntax errors) which are supported when void is generalized.

- `generalized_void_usage_test.dart`: This test uses a `void` typed expression in the 6 situations where an expression of type void must be accepted when void is generalized.

- `void_type_callbacks_test.dart`: This test contains expressions of type `void` in various contexts, along with test outcomes like `compile-time error` when needed for the new semantics.

- `void_type_function_types_test.dart`: Tests function types with `void` also as a parameter type.

- `void_type_override_test.dart`: Contains tests involving the `void` type and overriding.

There are other tests involving `void`, of course, but I could not find any others that are testing the generalized void feature. One test is obsolete in the sense that it should _fail_ under the new semantics: `void_check_test.dart`. The problem is that it contains the declaration 

```dart
callFoo(A a) => a.foo();
```

where `a.foo()` is an expression of type `void`, and that should incur a compile-time error. I suppose the best approach to fix this test would be to change the declaration to

```dart
callFoo(A a) => a.foo() as dynamic;
```

in which case it should still run successfully under the new semantics.

**Comment by eernstg** [comment]
I believe that the most productive approach, for any given tool/configuration, would be to make `generalized_void_syntax_test.dart` run (maybe it runs fine already) and then concentrate on `void_type_usage_test.dart`. The latter contains 239 multi-test comments with `compile-time error` and 36 with `ok`, so it touches on a lot of different contexts where `void` and void-typed expressions can occur.

**Comment by peter-ahe-google** [comment]
@eernstg thank you, I'll give that a try.

**Comment by eernstg** [comment]
@peter-ahe-google Thank you! Note also #31883 which explains why some subtests in `void_type_usage_test.dart` can be somewhat misleading. That shouldn't matter much, though, because there are so many other subtests in that test.

**Comment by peter-ahe-google** [comment]
Thank you, @a-siva. Unfortunately, work still remains on producing the correct compile-time errors.

**Comment by dgrove** [comment]
Since we're still using the analyzer to capture these compile-time errors, WDYT about moving this to the Dart 2-Beta 4 milestone, @a-siva ?

**Comment by mit-mit** [comment]
cc @peter-ahe-google to determine if we are OK to postpone this as analyzer catches it

**Comment by peter-ahe-google** [comment]
There's a lot of missing compile-time errors in both Fasta and Analyzer. This indicates to me that it would be a good idea to keep this in Dart2Stable, and I'm raising the priority accordingly.

**Comment by mit-mit** [comment]
@peter-ahe-google can you help us understand roughly how many missing errors we have, and what the rough estimate is for adding those?

**Comment by peter-ahe-google** [comment]
I'm assuming that all tests related to generalized `void` are in tests/language_2 or a subdirectory and have "void" in their name.

If so, there are 331 tests of generalized `void`. I have all them them passing except for 6 tests that are all related to the interaction of `FutureOr` and `void`. I think this boils down to implementing "invalid returns" for asynchronous functions (see https://dart-review.googlesource.com/c/sdk/+/60401/6/docs/language/informal/invalid_returns.md#46).

**Comment by peter-ahe-google** [comment]
I have one remaining test that is failing (tests/language_2/void/return_future_or_void_sync_error3_test.dart). I believe this corresponds to "invalid returns" for synchronous functions (see https://dart-review.googlesource.com/c/sdk/+/60401/6/docs/language/informal/invalid_returns.md#24). I should be able to have this working tomorrow.

**Comment by dgrove** [comment]
Any updates on this, @peter-ahe-google ?

**Comment by peter-ahe-google** [comment]
Update: instead of only running tests with "void" in their name (331 tests in total), I started running all tests (16019 tests in total if you don't count skipped tests). I started out having 319 of the 16019 tests failing. And I'm now down to 40 and have a better understanding of what is left:

* Rejecting void in ?? operator.
* Rejecting tests/language_2/void/return_future_or_void_sync_error3_test.dart

Unfortunately, the "specification" is being updated concurrently, so I'm not sure where this stands.

**Comment by peter-ahe-google** [comment]
Proposed fix: https://dart-review.googlesource.com/c/sdk/+/65141

I have one remaining issue that depends on #32613 which is in the process of being landed.

**Comment by peter-ahe-google** [comment]
Blocked on dart-lang/dartdoc#1724.

**Comment by peter-ahe-google** [comment]
The implementation was reverted.

**Comment by peter-ahe-google** [comment]
I've made all the tests work as written by the language team, yet there's still no finalized specification of this feature. Given that, I'm unable to make any progress on this.

**Comment by peter-ahe-google** [comment]
I'm taking a leap of faith and hoping that CL [66161](https://dart-review.googlesource.com/c/sdk/+/66161) will fix this and #33934.

**Comment by leafpetersen** [comment]
My understanding is the following:
  1) Certain parts of the proposed spec were too breaking to existing packages (notably mockito).  I don't think google3 was the primary problem.
  2) A few things got inadvertently missed in the analyzer implementation.  We may (or may not) be able to fix these at this point, depending on how breaking they are.

For 1), I believe that void -> void data flow had to be allowed, and also I believe allowing `dynamic foo() => voidValue`.

For 2), errata that I know about are:
  - `await voidValue` is allowed
     - @MichaelRFairhurst may have already landed a fix for this?
 - `if (voidValue)` is allowed
     - May be able to fix in future depending on how breaking

@MichaelRFairhurst can you comment on the delta from the spec in what you implemented per 1)?

@MichaelRFairhurst can you comment on any additional errata that you know about per 2)?

I attempted to rationalize, spec, test, and clean up the errors around returns as best I could.  The proposed spec for that is here: https://dart-review.googlesource.com/c/sdk/+/60401 , but it has not been signed off on.  That spec CL does reflect what I implemented in the analyzer, and what I landed tests for under tests/language_2/void .

**Comment by peter-ahe-google** [comment]
This is blocked on #33958.

**Comment by MichaelRFairhurst** [comment]
Sorry I didn't see this yesterday.

I wasn't aware of the `if (voidValue)` case, and that makes me sad. I imagine that should be a small change to make?

The other one I know of is `!voidValue` which also makes me sad and I would imagine would also be a small change.

I could see these both being dart 2.1 since they are that small. However, that makes me wonder if we shouldn't just go forward with them right now while they're on the mind.

There are two things that made this big enough to warrant a rollback, and I bet if they were fixed that any other issues would be edge cases we can iterate on:

- void assignment (this makes mockito's API insufficient for void methods)
- the popular package:coveralls triggers the edgecase `Future f() async { return void; }`, breaking lots of downstream packages

I *think* with these two things fixed it would be perfectly landable.

**Comment by MichaelRFairhurst** [comment]
> Sorry, I was mis-remembering. if (voidValue) is currently caught, it's !voidValue which is erroneously not caught.

Ah, and I screwed up my test confirming this...

Regarding

> await voidValue is allowed @MichaelRFairhurst may have already landed a fix for this?

That is submitting now.

**Comment by leafpetersen** [comment]
Since the presentation in terms of what returns are valid, I added an alternative presentation of the invalid returns proposal in the bottom half here:

https://dart.googlesource.com/sdk/+/refs/changes/01/60401/9/docs/language/informal/invalid_returns.md

See the "Alternative presentation" section at the bottom for what should be an equivalent set of rules presented in terms of what things are errors rather than what things are valid.

**Comment by vsmenon** [comment]
Pushing this out to 2.1.  We'll have the error in analyzer, but not CFE.  Not ideal, but that is already true for other errors.

**Comment by leafpetersen** [comment]
In progress CL sorting out language tests here: https://dart-review.googlesource.com/c/sdk/+/66521 for reference.  Tests under language_2/void have been triaged.

**Comment by leafpetersen** [comment]
tests have landed in https://dart-review.googlesource.com/c/sdk/+/66521

**Comment by leafpetersen** [comment]
I'm still seeing CFE errors on void to void transfers, and this seems to be causing trouble with rolling flutter into fuschia.  This example produces an incorrect error:

```dart
void foo() {}

void main () {
  void v = foo();
}
```

The tests I landed don't seem to have been triaged yet.

cc @tvolkert

**Comment by tvolkert** [comment]
FYI, this is producing spurious warnings in the Fuchsia build (it causes issues with mockito's `verify()` call), but it's not blocking a roll.

**Comment by leafpetersen** [comment]
FYI: I made a mistake while revising the spec for void related returns that implied an undesired breaking change.  I fixed the spec and the analyzer implementation, and adjusted the tests here: https://dart-review.googlesource.com/c/sdk/+/70462 .

**Comment by MichaelRFairhurst** [comment]
another user hitting this: https://github.com/dart-lang/mockito/issues/163

**Comment by Hixie** [comment]
For planning purposes, what's the status of this issue? Is a revert of https://dart-review.googlesource.com/c/sdk/+/66160 likely? Is the analyzer likely to change? Are the compiler warnings here to stay? Is the status quo likely to remain for the foreseeable future, or is something likely to change in the near term?

**Comment by kmillikin** [comment]
Some of the compiler warnings should be errors, and some of them are bugs in the implementation (they are false positive warnings where there is no problem).

https://github.com/dart-lang/sdk/commit/2a36502b028e435de547276631f91780981a5082 fixed many of the false positive warnings, but not all of them.  They were left as warnings.  [This CL](https://dart-review.googlesource.com/c/sdk/+/70862) should fix the remaining false positives and turn the genuine error cases back into static errors.

I will ensure that I verify that the errors do not break anything in Flutter that I can see, and ensure the behavior is as intended according to the language team, before landing the fix.

**Comment by Hixie** [comment]
Thanks. Will there be an effort to make sure the analyzer reports at least all the errors that the compiler reports? One of the weird things we're dealing with right now is that the compiler complains about things that the analyzer is happy with.

**Comment by MichaelRFairhurst** [comment]
@Hixie 

Yes, and I'll be the one to do that.

From what I know, the missing errors in the analyzer here are extreme edge cases (like `!print(...)`). So its not currently my top priority but I can see about fixing them sooner, for the sake of making this situation less confusing.

**Comment by Hixie** [comment]
Great! We see a lot of these messages when running our tests (from mockito, I assume), so we are looking forward to these changes making those go away. Let me know if there's anything we can do to help.

**Comment by kmillikin** [comment]
@leafpetersen @eernstg 

By my understanding of https://github.com/dart-lang/sdk/blob/master/docs/language/informal/invalid_returns.md this should be invalid:

```
FutureOr<void> f() => print('Hello');
```

It's a synchronous function and the return type is not `void` so we need to have that the expression would be a valid return expression in a block-bodied synchronous function.  None of the three cases for valid returns of an expression from a block-bodied synchronous function make this valid.

Is that correct?

**Comment by eernstg** [comment]
Discussing this IRL, Kevin and I concluded that [invalid_returns.md](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/invalid_returns.md) makes it an error: We go from 'Expression bodied functions', synchronous, to 'Block bodied functions', 'Synchronous', with `T == FutureOr<void>` and `S == void`, which fails to match any of the specified cases.

In the case where we decide that it "should" have been a valid return, I believe the relevant adjustment would be to require '`S` is voidy' rather than '`S` is `void`', where a type is voidy iff it is `void` or it is of the form `FutureOr<S2>` where `S2` is voidy. We could also say '`S` is `FutureOr^k<void>` for some `k >= 0`'.

**Comment by lrhn** [comment]
I agree with the interpretation. We did not go for "voidy" in the asynchronous cases either, and I'm not sure we should for the synchronous ones alone. This is a pure `void` expression occurring in a position where the value is going to be used (to check if it's a future). There is no need to make the return type `FutureOr<void>` in this case, `void` is sufficient and the function type will then be a proper subtype of the function type with `FutureOr<void>` as return type.

So, all in all, I'm not sure we need to change anything.

**Comment by kmillikin** [comment]
Something seems off about that.  I believe it's OK to write this (Dart 1 programmers do it all the time):

```dart
Future<int> future;
...future.then((i) => print('done'))...
```

If I notice that this same function is used multiple times, I might name it as a local function and I might declare a type for it.  (In fact, I will declare a type for it because I want the type checker to validate my understanding and I want the type to serve as documentation.)  I will just use the inferred type which is `FutureOr<void> Function(int)`.  So:

```dart
FutureOr<void> done(int i) => print('done');
...future.then(done)...
```

Here the implementation will tell me that I have a static error which is an invalid return.  But something's fishy about that, because all I did was move some code around and annotate it with its inferred type.  So either (1) there's nothing wrong with this return or (2) there is, but not both.

**Comment by kmillikin** [comment]
@Hixie the current status is that fixing the remaining false positive warnings for `void` requires us to have an implementation of the spec for invalid returns.  And implementing that specification (as specified) will give us even more false positive errors (which is undesirable).  So we need to sort out how to avoid those errors in a consistent way.

ETA is hopefully the end of this week.  There doesn't seem to be any implementation problems to sort out.

**Comment by kmillikin** [comment]
The front end is now passing the language_2 void and invalid returns tests.  This was landed in 9bdf2489b54841567005eaff7033f3451a047b37.

**Comment by leafpetersen** [comment]
> I will just use the inferred type which is FutureOr<void> Function(int). So:

Well, actually the inferred type is `void Function(int)` since we use the principal type for functions, and if you use that type, it's ok (this is why there wasn't an error in the lambda to start with).  But agreed that it's not obvious to the user that this is the inferred type.

These voidy warnings were something of a rabbit hole, so I'm hesitant to revisit them without some evidence that there is a real problem here.

**Comment by kmillikin** [comment]
Until a few days ago the inferred type was `FutureOr<void> Function(int)` but that was a bug and I fixed it.  Now we do have `void Function(int)`:

```dart
static method main() → dynamic {
  asy::Future<core::int> future;
  future.{asy::Future::then}<void>((core::int i) → void => core::print("done"));
}
```
